File: cpp/ProductListModel.cpp
#include "ProductListModel.h"
#include <QString>

ProductListModel::ProductListModel(const QVector<FinancialProduct*>& products, QObject* parent)
    : QAbstractTableModel(parent), m_products(products)
{
}

int ProductListModel::rowCount(const QModelIndex&) const
{
    return m_products.size();
}

int ProductListModel::columnCount(const QModelIndex&) const
{
    return 2;
}

QVariant ProductListModel::data(const QModelIndex& index, int role) const
{
    if (!index.isValid())
        return QVariant();

    FinancialProduct* product = m_products[index.row()];
    
    if (role == Qt::DisplayRole || role == Qt::EditRole) {
        switch (index.column()) {
            case 0: return product->name();
            case 1: return QString::number(product->price(), 'f', 2);
        }
    }
    return QVariant();
}

QVariant ProductListModel::headerData(int section, Qt::Orientation orientation, int role) const
{
    if (role == Qt::DisplayRole && orientation == Qt::Horizontal) {
        static const QStringList headers = {"Product Name", "Current Price"};
        return section < headers.size() ? headers[section] : QVariant();
    }
    return QVariant();
}

Qt::ItemFlags ProductListModel::flags(const QModelIndex& index) const
{
    Qt::ItemFlags flags = QAbstractTableModel::flags(index);
    if (index.isValid()) {
        flags |= Qt::ItemIsEnabled | Qt::ItemIsSelectable;
    }
    return flags;
} 

File: py/product_list_model.py
from PyQt6.QtCore import QAbstractTableModel, Qt, QModelIndex
from database_operations import load_products, save_product

class ProductListModel(QAbstractTableModel):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.products = load_products()

    def rowCount(self, parent=QModelIndex()):
        return len(self.products)

    def columnCount(self, parent=QModelIndex()):
        return 2  # Name and Price

    def data(self, index, role=Qt.ItemDataRole.DisplayRole):
        if not index.isValid():
            return None
            
        product = self.products[index.row()]
        
        if role in (Qt.ItemDataRole.DisplayRole, Qt.ItemDataRole.EditRole):
            if index.column() == 0:
                return product.name
            elif index.column() == 1:
                return f"{product.price:.2f}"
        return None

    def headerData(self, section, orientation, role=Qt.ItemDataRole.DisplayRole):
        if role == Qt.ItemDataRole.DisplayRole and orientation == Qt.Orientation.Horizontal:
            return ["Product Name", "Current Price"][section]
        return None

    def flags(self, index):
        flags = super().flags(index)
        if index.isValid():
            flags |= Qt.ItemFlag.ItemIsEnabled | Qt.ItemFlag.ItemIsSelectable
        return flags 

    def save_to_database(self):
        """Save all products to the database."""
        for product in self.products:
            # Implement saving logic for each product type
            pass 